executor/index_merge_reader.go

```diff
@@ -121,14 +128,27 @@ func (e *IndexMergeReaderExecutor) Open(ctx context.Context) error {
 			}
 			continue
 		}
-		keyRange, err := distsql.IndexRangesToKVRanges(e.ctx.GetSessionVars().StmtCtx, getPhysicalTableID(e.table), e.indexes[i].ID, e.ranges[i], e.feedbacks[i])
+		var pid int64
+		if is.Index.Global {
+			e.hasGlobalIndex = true
+			pid = e.table.Meta().ID
+		} else {
+			pid = getPhysicalTableID(e.table)
+		}
+		keyRange, err := distsql.IndexRangesToKVRanges(e.ctx.GetSessionVars().StmtCtx, pid, e.indexes[i].ID, e.ranges[i], e.feedbacks[i])
 		if err != nil {
 			return err
 		}
 		e.keyRanges = append(e.keyRanges, keyRange)
 	}
 	e.finished = make(chan struct{})
 	e.resultCh = make(chan *lookupTableTask, atomic.LoadInt32(&LookupTableTaskChannelSize))
+	if e.hasGlobalIndex && !e.skipGlobalIndex {
+		e.handleMaps = make(map[int64]*kv.HandleMap)
+		for _, p := range e.table.Meta().GetPartitionInfo().Definitions {
+			e.handleMaps[p.ID] = kv.NewHandleMap()
```

Comment:
But if we make the may lazy initialized, maybe there would be a data race

Reply:
I found that handles are unique in whole partitioned table. How about use a single global `handleMap` to store all handles and not clear when call `nextPartition`. But memory resource can't immediately recycle after processed every partitions.

Reply:
Or use two handle maps, one store handles read from global index, another store handles read from local index, so that local handle map could free after process every partition. But we must check twice for every handles.